src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    /// <summary>Provides a value type string builder composed of individual segments represented as <see cref="ReadOnlyMemory{T}"/> instances.</summary>
+    [DebuggerDisplay("Count = {_count}")]
+    internal struct SegmentStringBuilder
+    {
+        /// <summary>The array backing the builder, obtained from <see cref="ArrayPool{T}.Shared"/>.</summary>
+        private ReadOnlyMemory<char>[] _array;
+        /// <summary>The number of items in <see cref="_array"/>, and thus also the next position in the array to be filled.</summary>
+        private int _count;
+
+        /// <summary>Initializes the builder.</summary>
+        /// <param name="capacity">The initial capacity of the builder.</param>
+        public SegmentStringBuilder(int capacity)
+        {
+            Debug.Assert(capacity > 0);
+            _array = ArrayPool<ReadOnlyMemory<char>>.Shared.Rent(capacity);
+            _count = 0;
+        }
+
+        /// <summary>Gets the number of segments added to the builder.</summary>
+        public int Count => _count;
+
+        /// <summary>Adds a segment to the builder.</summary>
+        /// <param name="segment">The segment.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add(ReadOnlyMemory<char> segment)
+        {
+            ReadOnlyMemory<char>[] array = _array;
+            int pos = _count;
+            if ((uint)pos < (uint)array.Length)
+            {
+                array[pos] = segment;
+                _count = pos + 1;
+            }
+            else
+            {
+                GrowAndAdd(segment);
+            }
+        }
+
+        /// <summary>Grows the builder to accomodate another segment.</summary>
+        /// <param name="segment"></param>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void GrowAndAdd(ReadOnlyMemory<char> segment)
+        {
+            ReadOnlyMemory<char>[] array = _array;
+            Debug.Assert(array.Length == _count);
+
+            ReadOnlyMemory<char>[] newArray = _array = ArrayPool<ReadOnlyMemory<char>>.Shared.Rent(array.Length * 2);
+            Array.Copy(array, newArray, _count);
+            ArrayPool<ReadOnlyMemory<char>>.Shared.Return(array, clearArray: true);
+            newArray[_count++] = segment;
+        }
+
+        /// <summary>Gets a span of all segments in the builder.</summary>
+        /// <returns></returns>
+        public Span<ReadOnlyMemory<char>> AsSpan() => new Span<ReadOnlyMemory<char>>(_array, 0, _count);
+
+        /// <summary>Creates a string from all the segments in the builder and then disposes of the builder.</summary>
```

Comment:
> and then disposes of the builder

Is this a typical pattern for `ToString()` to mutate/dispose the instance? I think I would find that behavior surprising if I used this type.

Reply:
On public types, probably not.  But it's what we do in ValueStringBuilder:
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L86-L91